src/catchup/CatchupManagerImpl.cpp

```diff
@@ -28,32 +28,132 @@ CatchupManager::create(Application& app)
 }
 
 CatchupManagerImpl::CatchupManagerImpl(Application& app)
-    : mApp(app), mCatchupWork(nullptr)
+    : mApp(app)
+    , mCatchupWork(nullptr)
+    , mSyncingLedgersSize(
+          app.getMetrics().NewCounter({"ledger", "memory", "queued-ledgers"}))
 {
 }
 
 CatchupManagerImpl::~CatchupManagerImpl()
 {
 }
 
+uint32_t
+CatchupManagerImpl::getCatchupCount()
+{
+    return mApp.getConfig().CATCHUP_COMPLETE
+               ? std::numeric_limits<uint32_t>::max()
+               : mApp.getConfig().CATCHUP_RECENT;
+}
+
 void
-CatchupManagerImpl::historyCaughtup()
+CatchupManagerImpl::processLedger(LedgerCloseData const& ledgerData)
 {
-    mCatchupWork.reset();
+    uint32_t lastReceivedLedgerSeq = ledgerData.getLedgerSeq();
+
+    // Caught up
+    if (lastReceivedLedgerSeq ==
+        mApp.getLedgerManager().getLastClosedLedgerNum())
+    {
+        reset();
+        return;
+    }
+
+    // We are not caught up, but catchup work is finished
+    if (mCatchupWork && mCatchupWork->isDone())
+    {
+        trimSyncingLedgers();
+        mCatchupWork.reset();
+    }
+
+    auto addLedgerData = [&]() { addToSyncingLedgers(ledgerData); };
+
+    if (!mCatchupWork)
+    {
+        uint32_t checkpointLedger;
+        if (!mSyncingLedgers.empty())
+        {
+            // front of mSyncingLedgers should always be a checkpoint ledger if
+            // catchup is not running
+            uint32_t firstBufferedLedgerSeq =
+                mSyncingLedgers.front().getLedgerSeq();
+
+            checkpointLedger = mApp.getHistoryManager().nextCheckpointLedger(
+                firstBufferedLedgerSeq - 1);
+            assert(checkpointLedger == firstBufferedLedgerSeq);
+        }
+        else
+        {
+            checkpointLedger = mApp.getHistoryManager().nextCheckpointLedger(
```

Comment:
Good catch! That logic was used to identify the ledger after a checkpoint because I originally didn't require the checkpoint ledger to be buffered (which was wrong).